### PR TITLE
Blog-Application: Add authorization rules 📐 

### PR DESCRIPTION
**Pull request by evansnyamekye**
- Install CanCanCan in your project.
- Add a role column to the user's table. Remember to use a migration for this.
A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
- [ ] For that you need to implement the post-deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
- [ ] For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

@evansnyamekye 
@Obimbo07 
@Izebiafe 

Thank you 👨👨👨

### CODE REVIEW

**Review by zeff96 — COMMENTED** [review]
## STATUS: INVALID CODE REVIEW 🛑
Hi @evansnyamekye, @Obimbo07 and @Izebiafe , 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that lacks equal contribution into the codebase from all partners, @Izebiafe kindly make sure to have fair contribution. Please go ahead and close this PR without merging and open a new PR with equal commits from all partners

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: app/controllers/posts_controller.rb
Comment (fobadara) [inline app/controllers/posts_controller.rb:1]: [REQUIRED] Kindly use CanCan for post authorization. You can do this:
```
class PostsController < ApplicationController
    load_and_authorize_resource

    ...
  end
```

**Review by fobadara — CHANGES_REQUESTED** [review]
Hi team, 

Good job so far!
You still need to work on some issues to go to the next project, but you are almost there!

REDO

## Highlights
✅ Descriptive README
✅ Descriptive PR summary
✅ Clean code
✅ Good first attempt
✅ Good implementation of the principle of separating concerns.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb
Comment (skyv26) [inline app/controllers/posts_controller.rb:1]: - [ ] Please adhere to the project requirements, your previous code reviewer asked for the adding the `callback` attribute `load_and_authorize_resource` to auto load the `authorization` for the posts. But it seems like you didn't add it, kindly add the required changes and re-submit your project again.

**Review by skyv26 — CHANGES_REQUESTED** [review]
## Hi @evansnyamekye, @fobadara and @Izebiafe, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but some are still not addressed yet.

## STATUS: REQUIRE CHANGES 🔴 



## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implemented the features ✅ 
- Nice *UI* design ✅

## Required Changes ♻️

- [ ] Dear Team members, you have done a great job so far, but it seems like you missed some core project requirements, according to which you should add the functionalities to delete `post` and `comments`. Only in two conditions, it should work, either a person who is authorized as an `admin` can delete any `post` and `comment` of any user **OR** a normal user can delete its own `posts` or `comments` on the blog. Right now, the add comments, like. delete button all these are missing in your project because it is a part of the requirement, which needs to be fulfilled to pass the milestone. Kindly take a look at the below screenshot and do things accordingly before asking for the next review.

image


_Check the comments under the review._


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

#### Cheers, and Happy coding!👏👏👏

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb
Comment (Obimbo07) [reply app/controllers/posts_controller.rb:1]: We Resolved the posts authorization

File: app/views/posts/show.html.erb [hunk 1]
```diff
@@ -3,12 +3,27 @@
...
-      <div class="comment">
-        <p><strong><%= comment.user.name %> :</strong> <%= comment.text %></p>
+      <div class="row align-items-end comment">
+        <div class="col">
+          <%= comment.user.name %>: <%= comment.text %>
+          <% if can? :destroy , comment %>
+            <%= form_with url: user_post_comment_path(@user, @post, comment), method: :delete do |form| %>
+              <%= form.submit 'Delete Comment', data: { confirm: 'Are you sure?' } %>
+            <% end %>
+          <% end %>
```
Comment (shahadat3669) [inline app/views/posts/show.html.erb:24]: - [ ] Great work so far! However, I noticed that there is an issue with the comment functionality. When I tried to delete a comment as a user or admin, it did not work as expected. I received an `ActiveRecord::RecordNotFound` error in the console for the comment. Additionally, I noticed that the code is checking for a post with ID 3 (while the post ID was 4) in line number 20 of your controller. So I was unable to test the authorization. To fix this issue, I recommend that you update your code and use the proper parameters for it.

File: app/controllers/posts_controller.rb
Comment (shahadat3669) [inline app/controllers/posts_controller.rb:1]: - [ ] Great job so far! Your post-delete controller is working properly. However, as the previous reviewer mentioned, please add 'load_and_authorize_resource' before all of your controller methods.

```ruby
 class PostsController < ApplicationController
    load_and_authorize_resource

    ...
  end
```

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi every one, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ No linter Errors
✅ Descriptive PR Title
✅ Descriptive PR Description



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -21,7 +21,15 @@
...
+        <% if can? :destroy , post %>
+          <%= form_with(url: user_post_path(@user, post), method: :delete) do |form| %>
+            <%= form.submit 'Delete post', data: { confirm: 'Are you sure?' }, class: "delete-button btn" %>
+          <% end %>
+        <% end %>
+      </div>
     </div>
+    
   <% end %>
   <%= link_to "See all posts",user_posts_path(@user), class: "button" %>
```
Comment (prolajumokeoni) [inline app/views/users/show.html.erb:34]: ### OPTIONAL
 
 - Good job so far, when I click the see all post I see this error ⬇️ 
Screenshot_243

- [ ] You can update your code to show a popup of `no posts yet` and add a button for users to create a new post,  on each user's profile 👍🏿

**Review by prolajumokeoni — APPROVED** [review]
Hi @Team

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
 - load_and_authorize_resource to post controller ✅ 
 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**